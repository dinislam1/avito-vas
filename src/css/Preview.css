.outer {
  display: flex;
  align-items: center;
  flex-direction: column;
  justify-content: center;
  flex: 1;
  flex-wrap: wrap;
}
