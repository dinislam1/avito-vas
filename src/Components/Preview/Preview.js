import React, { useEffect } from "react";
import { fabric } from "fabric";
import { ColorLuminance } from "./PreviewFunctions";
import "css/Preview.css";

let canvas = null;
let rect = null;
let textbox = null;
const [width, height] = [420, 300];

export function Preview(state) {
  let { bgColor, fontSize, fontFamily, pastedURL, text } = state.init || {};

  useEffect(() => {
    canvas = new fabric.Canvas(document.getElementById("canvas"), {});
    rect = new fabric.Rect({
      selectable: false,
      hoverCursor: "default",
    });
    textbox = new fabric.Text("Avito", {
      left: canvas.width / 3,
      top: canvas.height / 4,
      width: canvas.width / 2,
      editable: false,
    });
  }, []);

  useEffect(() => {
    rect.width = width;
    rect.height = height;
    textbox.fontSize = fontSize;
    textbox.text = text;
    textbox.fontFamily = fontFamily;
    rect.setGradient("fill", {
      x1: 0,
      y1: 0,
      x2: rect.width,
      y2: rect.height,
      colorStops: {
        0: ColorLuminance(bgColor.hex, -0.7),
        0.5: bgColor.hex,
        1: ColorLuminance(bgColor.hex, 0.7),
      },
    });

    fabric.util.loadImage(
      pastedURL,
      (img) => {
        if (img == null) {
          alert("Ошибка CORS либо ссылка на файл длинная, попробуйте заново");
          return;
        }

        const image = new fabric.Image(img);
        image.set({ left: 25, top: 75 }).scale(0.6);
        let objs = canvas.getObjects();
        if (objs.length) {
          objs.forEach((e) => {
            if (e && e.type === "image") {
              e._element.src = pastedURL;
              canvas.renderAll();
            }
          });
        } else {
          canvas.add(rect);
          canvas.add(image).setActiveObject(image);
          canvas.add(textbox);
        }
      },
      null,
      "anonymous"
    );
    canvas.renderAll();
  }, [bgColor, pastedURL, fontSize, fontFamily, text]);

  return <canvas id="canvas" width={width} height={height} />;
}
