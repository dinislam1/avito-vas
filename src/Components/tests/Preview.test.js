import React from "react";
import { shallow, mount, render } from "./enzyme";
import { Preview } from "Components/Preview/Preview";
import renderer from "react-test-renderer";
import "jest-canvas-mock";

const setUp = (props) => shallow(<Preview {...props} />);

describe("Preview Test Suite", () => {
  let component;
  beforeEach(() => {
    component = setUp();
  });

  it("should contain canvas", () => {
    const wrapper = component.find("#canvas");
    expect(wrapper.length).toBe(1);
  });
});

describe("Preview component", () => {
  it("should render Preview component", () => {
    const tree = renderer.create(<Preview />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
