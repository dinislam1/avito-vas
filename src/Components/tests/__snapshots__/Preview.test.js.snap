// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preview component should render Preview component 1`] = `
<canvas
  height={300}
  id="canvas"
  width={420}
/>
`;
